src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1040 @@
+
+       -- generate pseudo-random values until we exceed the target magnitude
+       f mag v g | mag >= magtgt = (v, g)
+                 | otherwise = v' `seq`f (mag*b) v' g' where
+                        (x,g') = next g
+                        v' = v * b + (fromIntegral x - fromIntegral genlo)
+
+-- | Generate an integral in the range @[l, h]@ if @l <= h@ and @[h, l]@
+-- otherwise.
+uniformIntegralM :: (Bits a, Integral a, StatefulGen g m) => (a, a) -> g -> m a
+uniformIntegralM (l, h) gen = case l `compare` h of
+  LT -> do
+    let limit = h - l
+    let limitAsWord64 :: Word64 = fromIntegral limit
+    bounded <-
+      if fromIntegral limitAsWord64 == limit
+        -- Optimisation: if 'limit' fits into 'Word64', generate a bounded
+        -- 'Word64' and then convert to 'Integer'
+        then fromIntegral <$> unsignedBitmaskWithRejectionM uniformWord64 limitAsWord64 gen
+        else boundedExclusiveIntegralM (limit + 1) gen
+    return $ l + bounded
+  GT -> uniformIntegralM (h, l) gen
+  EQ -> pure l
+{-# INLINEABLE uniformIntegralM #-}
+
+-- | Generate an integral in the range @[0, s)@ using a variant of Lemire's
+-- multiplication method.
+--
+-- Daniel Lemire. 2019. Fast Random Integer Generation in an Interval. In ACM
+-- Transactions on Modeling and Computer Simulation
+-- https://doi.org/10.1145/3230636
+--
+-- PRECONDITION (unchecked): s > 0
+boundedExclusiveIntegralM :: forall a g m . (Bits a, Integral a, StatefulGen g m) => a -> g -> m a
+boundedExclusiveIntegralM (s :: a) gen = go
+  where
+    n = integralWordSize s
+    -- We renamed 'L' from the paper to 'k' here because 'L' is not a valid
+    -- variable name in Haskell and 'l' is already used in the algorithm.
+    k = wordSizeInBits * n
+    twoToK = (1::a) `shiftL` k
+    modTwoToKMask = twoToK - 1
+
+    t = (twoToK - s) `mod` s
+    go :: (Bits a, Integral a, StatefulGen g m) => m a
+    go = do
+      x <- uniformIntegralWords n gen
+      let m = x * s
+      -- m .&. modTwoToKMask == m `mod` twoToK
+      let l = m .&. modTwoToKMask
+      if l < t
+        then go
+        -- m `shiftR` k == m `quot` twoToK
+        else return $ m `shiftR` k
+{-# INLINE boundedExclusiveIntegralM #-}
+
+-- | @integralWordSize i@ returns that least @w@ such that
+-- @i <= WORD_SIZE_IN_BITS^w@.
+integralWordSize :: (Bits a, Num a) => a -> Int
+integralWordSize = go 0
+  where
+    go !acc i
+      | i == 0 = acc
+      | otherwise = go (acc + 1) (i `shiftR` wordSizeInBits)
+{-# INLINE integralWordSize #-}
+
+-- | @uniformIntegralWords n@ is a uniformly pseudo-random integral in the range
+-- @[0, WORD_SIZE_IN_BITS^n)@.
+uniformIntegralWords :: (Bits a, Integral a, StatefulGen g m) => Int -> g -> m a
+uniformIntegralWords n gen = go 0 n
+  where
+    go !acc i
+      | i == 0 = return acc
+      | otherwise = do
+        (w :: Word) <- uniformM gen
+        go ((acc `shiftL` wordSizeInBits) .|. fromIntegral w) (i - 1)
+{-# INLINE uniformIntegralWords #-}
```

Comment:
If `integer-gmp` is available, one can fill a random bytearray and invoke [importIntegerFromByteArray](http://hackage.haskell.org/package/integer-gmp-1.0.3.0/docs/GHC-Integer-GMP-Internals.html#v:importIntegerFromByteArray).

Reply:
This is also a very nice idea. Let's create a separate ticket and we can tackle further `Integer` and `Natural` optimizations in the next release. What do you think?

Reply:
I am fine with splitting that work out, shouldn't be a blocker on this I don't think

Reply:
Filed here: https://github.com/idontgetoutmuch/random/issues/171 (on idontgetoutmuch/random for now, my idea was that we'd migrate outstanding issues to haskell/random once #62 is merged.)